README.md
>* NOTE: This package is a continuation of [data_connection_checker](https://github.com/komapeb/data_connection_checker) which currently is not continued. * 

# 🌍 Internet Connection Checker

[![Dart CI](https://github.com/RounakTadvi/internet_connection_checker/actions/workflows/main.yaml/badge.svg)](https://github.com/RounakTadvi/internet_connection_checker/actions/workflows/main.yaml)
[![codecov](https://codecov.io/gh/RounakTadvi/internet_connection_checker/branch/main/graph/badge.svg)](https://codecov.io/gh/RounakTadvi/internet_connection_checker)
[![style: very good analysis](https://img.shields.io/badge/style-very_good_analysis-B22C89.svg)](https://pub.dev/packages/very_good_analysis)
[![License: MIT](https://img.shields.io/badge/license-MIT-purple.svg)](https://opensource.org/licenses/MIT)

A Pure Dart Utility library that checks for an Active Internet connection by opening a socket to a list of specified addresses, each with individual port and timeout. Defaults are provided for convenience.

>*Note that this plugin is in beta and may still have
a few issues. [Feedback][issues_tracker] is welcome.*

### Table of contents

- [🌍 Internet Connection Checker](#-internet-connection-checker)
    - [Table of contents](#table-of-contents)
  - [Description](#description)
  - [Quick start](#quick-start)
  - [Purpose](#purpose)
  - [How it works](#how-it-works)
  - [Defaults](#defaults)
      - [`DEFAULT_ADDRESSES`](#default_addresses)
      - [`DEFAULT_PORT`](#default_port)
      - [`DEFAULT_TIMEOUT`](#default_timeout)
      - [`DEFAULT_INTERVAL`](#default_interval)
  - [Usage](#usage)
    - [Singleton example](#singleton-example)
    - [Create instance example](#create-instance-example)
  - [Features and bugs](#features-and-bugs)

## Description

Checks for an internet (data) connection, by opening a socket to a list of addresses.

The defaults of the plugin should be sufficient to reliably determine if
the device is currently connected to the global network, e.i. has access to the Internet.

>Note that you should not be using the current network status for deciding whether you can reliably make a network connection. Always guard your app code against timeouts and errors that might come from the network layer.

## Quick start

`InternetConnectionChecker()` is actually a Singleton. Calling `InternetConnectionChecker()`
is guaranteed to always return the same instance.

You can supply a new list to `InternetConnectionChecker().addresses` if you
need to check different destinations, ports and timeouts.
Also, each address can have its own port and timeout.
See `InternetAddressCheckOptions` in the docs for more info.

***First you need to [install it][install] (this is the preferred way)***

Then you can start using the library:

```dart
bool result = await InternetConnectionChecker().hasConnection;
if(result == true) {
  print('YAY! Free cute dog pics!');
} else {
  print('No internet :( Reason:');
  print(InternetConnectionChecker().lastTryResults);
}
```

## Purpose

The reason this package exists is that `connectivity_plus` package cannot reliably determine if a data connection is actually available. More info on its page here: <https://pub.dev/packages/connectivity_plus>

More info on the issue in general:

- <https://stackoverflow.com/questions/1560788/how-to-check-internet-access-on-android-inetaddress-never-times-out/27312494#27312494> (this is the best approach so far IMO and it's what I'm using)

You can use this package in combination with `connectivity_plus` in the following way:

```dart
var isDeviceConnected = false;

var subscription = Connectivity().onConnectivityChanged.listen((ConnectivityResult result) async {
  if(result != ConnectivityResult.none) {
    isDeviceConnected = await InternetConnectionChecker().hasConnection;
  }
});
```

*Note: remember to properly cancel the `subscription` when it's no longer needed. See `connectivity_plus` package docs for more info.*

## How it works

All addresses are pinged simultaneously. On successful result (socket connection to address/port succeeds) a `true` boolean is pushed to a list, on failure (usually on timeout, default 10 sec) a `false` boolean is pushed to the same list.

When all the requests complete with either success or failure, a check is made to see if the list contains at least one `true` boolean. If it does, then an external address is available, so we have data connection. If all the values in this list are `false`, then we have no connection to the outside world of cute cat and dog pictures, so `hasConnection` also returns `false` too.

This all happens at the same time for all addresses, so the maximum waiting time is the address with the highest specified timeout, in case it's unreachable.

I believe this is a ***reliable*** and ***fast*** method to check if a data connection is available to a device, but I may be wrong. I suggest you open an issue on the Github repository page if you have a better way of.

## Defaults

The defaults are based on data collected from <https://perfops.net/>, <https://www.dnsperf.com/#!dns-resolvers>

Here's some more info about the defaults:

#### `DEFAULT_ADDRESSES`

... includes the top 3 globally available free DNS resolvers.

| Address        | Provider   | Info                                              |
| :------------- | :--------- | :------------------------------------------------ |
| 1.1.1.1        | CloudFlare | <https://1.1.1.1>                                 |
| 1.0.0.1        | CloudFlare | <https://1.1.1.1>                                 |
| 8.8.8.8        | Google     | <https://developers.google.com/speed/public-dns/> |
| 8.8.4.4        | Google     | <https://developers.google.com/speed/public-dns/> |
| 208.67.222.222 | OpenDNS    | <https://use.opendns.com/>                        |
| 208.67.220.220 | OpenDNS    | <https://use.opendns.com/>                        |

```dart
static final List<AddressCheckOptions> DEFAULT_ADDRESSES =
      List<AddressCheckOptions>.unmodifiable(
    <AddressCheckOptions>[
      AddressCheckOptions(
        address: InternetAddress(
          '1.1.1.1', // CloudFlare
          type: InternetAddressType.IPv4,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2606:4700:4700::1111', // CloudFlare
          type: InternetAddressType.IPv6,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '8.8.4.4', // Google
          type: InternetAddressType.IPv4,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2001:4860:4860::8888', // Google
          type: InternetAddressType.IPv6,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '208.67.222.222', // OpenDNS
          type: InternetAddressType.IPv4,
        ), // OpenDNS
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2620:0:ccc::2', // OpenDNS
          type: InternetAddressType.IPv6,
        ), // OpenDNS
      ),
    ],
  );
```

#### `DEFAULT_PORT`

... is 53.

>A DNS server listens for requests on port 53 (both UDP and TCP). So all DNS requests are sent to port 53 ...

More info:

- <https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers>
- <https://www.google.com/search?q=dns+server+port>

```dart
static const int DEFAULT_PORT = 53;
```

#### `DEFAULT_TIMEOUT`

... is 10 seconds.

```dart
static const Duration DEFAULT_TIMEOUT = Duration(seconds: 10);
```

#### `DEFAULT_INTERVAL`

... is 10 seconds. Interval is the time between automatic checks. Automatic
checks start if there's a listener attached to `onStatusChange`, thus remember
to cancel unneeded subscriptions.

`checkInterval` (which controls how often a check is made) defaults
to this value. You can change it if you need to perform checks more often
or otherwise.

```dart
static const Duration DEFAULT_INTERVAL = const Duration(seconds: 10);
...
Duration checkInterval = DEFAULT_INTERVAL;
```

## Usage

The `InternetConnectionChecker` can be used as a singleton or can be instantiated with custom values.

### Singleton example

```dart
import 'package:internet_connection_checker/internet_connection_checker.dart';

main() async {
  // Simple check to see if we have internet
  print("The statement 'this machine is connected to the Internet' is: ");
  print(await InternetConnectionChecker().hasConnection);
  // returns a bool

  // We can also get an enum value instead of a bool
  print("Current status: ${await InternetConnectionChecker().connectionStatus}");
  // prints either InternetConnectionStatus.connected
  // or InternetConnectionStatus.disconnected

  // This returns the last results from the last call
  // to either hasConnection or connectionStatus
  print("Last results: ${InternetConnectionChecker().lastTryResults}");

  // actively listen for status updates
  // this will cause InternetConnectionChecker to check periodically
  // with the interval specified in InternetConnectionChecker().checkInterval
  // until listener.cancel() is called
  var listener = InternetConnectionChecker().onStatusChange.listen((status) {
    switch (status) {
      case InternetConnectionStatus.connected:
        print('Data connection is available.');
        break;
      case InternetConnectionStatus.disconnected:
        print('You are disconnected from the internet.');
        break;
    }
  });

  // close listener after 30 seconds, so the program doesn't run forever
  await Future.delayed(Duration(seconds: 30));
  await listener.cancel();
}
```

*Note: Remember to dispose of any listeners,
when they're not needed to prevent memory leaks,
e.g. in a* `StatefulWidget`'s *dispose() method*:
  
```dart
...
@override
void dispose() {
  listener.cancel();
  super.dispose();
}
...
```

See `example` folder for more examples.

### Create instance example

```dart
import 'package:internet_connection_checker/internet_connection_checker.dart';

main() async {
  final customInstance = InternetConnectionChecker.createInstance(
    checkTimeout: const Duration(seconds: 1), // Custom check timeout
    checkInterval: const Duration(seconds: 1), // Custom check interval
    addresses: [
      ... // Custom addresses
    ],
  );

  // Register it with any dependency injection framework. For example GetIt.
  GetIt.registerSingleton<InternetConnectionChecker>(
    customInstance,
  );
}
```

*Note: Remember to dispose of any listeners,
when they're not needed to prevent memory leaks,
e.g. in a* `StatefulWidget`'s *dispose() method*:
  
```dart
...
@override
void dispose() {
  listener.cancel();
  super.dispose();
}
...
```

See `example` folder for more examples.

## Features and bugs

Please file feature requests and bugs at the [issue tracker][issues_tracker].

[issues_tracker]: https://github.com/RounakTadvi/internet_connection_checker/issues
[pull_requests]: https://github.com/RounakTadvi/internet_connection_checker/pulls


CHANGELOG.md
## 1.0.0+1

- Updated Documentation (README.md)

## 1.0.0

- **Breaking change**: Ability to provide a host name instead of IP address. Thanks [gampixi](https://github.com/gampixi)!. Update the AddressCheckOptions constructor if custom addresses are being used since the positional argument address is now a named, optional argument.
- Update dependencies to the latest release
  
## 0.0.1+4

- Add option to customize the interval, timeout, and network addresses. Thanks [Jop Middelkamp](https://github.com/jopmiddelkamp)!

## 0.0.1+3

- Add IPv6 hosts to the default addresses. Thanks [MrCsabaToth](https://github.com/MrCsabaToth)!
- Update dependencies to the latest release
- Documentation (minor documentation updates). Thanks [AleksandarSavic95](https://github.com/AleksandarSavic95)!
- Code refactor

## 0.0.1+2

- Reports that an internet connection is available as soon as one request succeeds instead of potentially waiting for the last to time out.
- Removed universal_io dependency

## 0.0.1+1

- Updated Documentation
- Update a dependency to the latest release.

## 0.0.1

- Initial Release (non-breaking change which adds functionality)
- Documentation (minor documentation updates)
- Code refactor


test/address_check_result_test.dart
import 'dart:io';

import 'package:internet_connection_checker/internet_connection_checker.dart';
import 'package:test/test.dart';

void main() {
  test(
    'should verify toString() method',
    () {
      // Setup - Arrange
      final InternetAddress tInternetAddress = InternetAddress('1.1.1.1');
      const int DEFAULT_PORT = 43;
      const bool isSuccess = true;
      final AddressCheckOptions tOptions = AddressCheckOptions(
        address: tInternetAddress,
        port: DEFAULT_PORT,
      );

      final AddressCheckResult tResult =
          AddressCheckResult(tOptions, isSuccess: isSuccess);
      // Action - Act

      // Result - Assert
      expect(tResult.toString(), 'AddressCheckResult($tOptions, $isSuccess)');
    },
  );
}


test/address_check_options_test.dart
import 'dart:io';

import 'package:internet_connection_checker/internet_connection_checker.dart';
import 'package:test/test.dart';

void main() {
  test(
    'should verify toString() method',
    () {
      // Setup - Arrange
      final InternetAddress tInternetAddress = InternetAddress('1.1.1.1');
      const int DEFAULT_PORT = 43;
      const Duration DEFAULT_TIMEOUT = Duration(seconds: 10);
      final AddressCheckOptions tOptions = AddressCheckOptions(
        address: tInternetAddress,
        port: DEFAULT_PORT,
      );
      // Action - Act

      // Result - Assert
      expect(
        tOptions.toString(),
        'AddressCheckOptions($tInternetAddress, $DEFAULT_PORT, $DEFAULT_TIMEOUT)',
      );
    },
  );

  test('should allow only address', () {
    // Setup - Arrange
    final options = AddressCheckOptions(
      address: InternetAddress('1.1.1.1'),
    );

    // Result - Assert
    // Will pass if arguments are correct, as otherwise will have thrown
    expect(true, equals(true));
  });

  test('should allow only hostname', () {
    // Setup - Arrange
    final options = AddressCheckOptions(
      hostname: 'google.com',
    );

    // Result - Assert
    // Will pass if arguments are correct, as otherwise will have thrown
    expect(true, equals(true));
  });

  test('should not allow no address or hostname', () {
    // Setup - Arrange
    expect(() => AddressCheckOptions(), throwsA(isA<AssertionError>()));
  });

  test('should not allow both address and hostname', () {
    // Setup - Arrange
    expect(
      () => AddressCheckOptions(
        address: InternetAddress('1.1.1.1'),
        hostname: 'google.com',
      ),
      throwsA(
        isA<AssertionError>(),
      ),
    );
  });
}


test/internet_connection_checker_test.dart
import 'dart:async';

import 'package:internet_connection_checker/internet_connection_checker.dart';
import 'package:test/test.dart';

void main() {
  group('internet_connection_checker', () {
    StreamSubscription<InternetConnectionStatus>? listener1;
    StreamSubscription<InternetConnectionStatus>? listener2;

    tearDown(() {
      // destroy any active listener after each test
      listener1?.cancel();
      listener2?.cancel();
    });

    test('''Shouldn't have any listeners attached''', () {
      expect(
        InternetConnectionChecker().hasListeners,
        isFalse,
      );
    });

    test('''Unawaited call hasConnection should return a Future<bool>''', () {
      expect(
        InternetConnectionChecker().hasConnection,
        isA<Future<bool>>(),
      );
    });

    test('''Awaited call to hasConnection should return a bool''', () async {
      expect(
        await InternetConnectionChecker().hasConnection,
        isA<bool>(),
      );
    });

    test(
        '''Unawaited call to connectionStatus '''
        '''should return a Future<InternetConnectionStatus>''', () {
      expect(
        InternetConnectionChecker().connectionStatus,
        isA<Future<InternetConnectionStatus>>(),
      );
    });

    test(
        '''Awaited call to connectionStatus '''
        '''should return a Future<InternetConnectionStatus>''', () async {
      expect(
        await InternetConnectionChecker().connectionStatus,
        isA<InternetConnectionStatus>(),
      );
    });

    test('''We shouldn't have any listeners 1''', () {
      expect(
        InternetConnectionChecker().hasListeners,
        isFalse,
      );
    });

    test('''We should have listeners 1''', () {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().hasListeners,
        isTrue,
      );
    });

    test('''We should have listeners 2''', () {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().hasListeners,
        isTrue,
      );
    });

    test('''We should have listeners 3''', () async {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener1!.cancel();
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().hasListeners,
        isTrue,
      );
    });

    test('''We shouldn't have any listeners 2''', () async {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener1!.cancel();
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener2!.cancel();
      expect(
        InternetConnectionChecker().hasListeners,
        isFalse,
      );
    });

    test('''We shouldn't have any listeners 1''', () {
      expect(
        InternetConnectionChecker().hasListeners,
        isFalse,
      );
    });

    test('''We should have listeners 1 [isActivelyChecking]''', () {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().isActivelyChecking,
        isTrue,
      );
    });

    test('''We should have listeners 2 [isActivelyChecking]''', () {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().isActivelyChecking,
        isTrue,
      );
    });

    test('''We should have listeners 3 [isActivelyChecking]''', () async {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener1!.cancel();
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      expect(
        InternetConnectionChecker().isActivelyChecking,
        isTrue,
      );
    });

    test('''We shouldn't have any listeners 2 [isActivelyChecking]''',
        () async {
      listener1 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener1!.cancel();
      listener2 = InternetConnectionChecker().onStatusChange.listen((_) {});
      await listener2!.cancel();
      expect(
        InternetConnectionChecker().isActivelyChecking,
        isFalse,
      );
    });

    test('''We should be able to set a custom timeout value''', () async {
      const Duration timeout = Duration(seconds: 1);
      final InternetConnectionChecker internetConnectionChecker =
          InternetConnectionChecker.createInstance(
        checkTimeout: timeout,
      );
      expect(
        internetConnectionChecker.addresses.every(
          (AddressCheckOptions element) => element.timeout == timeout,
        ),
        isTrue,
      );
    });

    test('''We should be able to set a custom interval value''', () async {
      const Duration interval = Duration(seconds: 1);
      final InternetConnectionChecker internetConnectionChecker =
          InternetConnectionChecker.createInstance(
        checkInterval: interval,
      );
      expect(
        internetConnectionChecker.checkInterval,
        interval,
      );
    });

    test('''We should be able to set a custom timeout value''', () async {
      const Duration timeout = Duration(seconds: 1);
      final InternetConnectionChecker internetConnectionChecker =
          InternetConnectionChecker.createInstance(
        checkTimeout: timeout,
      );
      expect(
        internetConnectionChecker.checkTimeout,
        timeout,
      );
      expect(
        internetConnectionChecker.addresses.every(
          (AddressCheckOptions element) => element.timeout == timeout,
        ),
        isTrue,
      );
    });

    test('''We should be able to set custom addresses''', () async {
      final List<AddressCheckOptions> addresses = <AddressCheckOptions>[
        InternetConnectionChecker.DEFAULT_ADDRESSES.first,
      ];
      final InternetConnectionChecker internetConnectionChecker =
          InternetConnectionChecker.createInstance(
        addresses: addresses,
      );
      expect(
        internetConnectionChecker.addresses,
        addresses,
      );
    });
  });
}


example/example.dart
import 'dart:async';

import 'package:internet_connection_checker/internet_connection_checker.dart';

Future<void> main() async {
  // Simple check to see if we have Internet
  // ignore: avoid_print
  print('The statement \'this machine is connected to the Internet\' is: ');
  final bool isConnected = await InternetConnectionChecker().hasConnection;
  // ignore: avoid_print
  print(
    isConnected.toString(),
  );
  // returns a bool

  // We can also get an enum instead of a bool
  // ignore: avoid_print
  print(
      'Current status: ${await InternetConnectionChecker().connectionStatus}');
  // Prints either InternetConnectionStatus.connected
  // or InternetConnectionStatus.disconnected

  // actively listen for status updates
  StreamSubscription<InternetConnectionStatus> listener =
      InternetConnectionChecker().onStatusChange.listen(
    (InternetConnectionStatus status) {
      switch (status) {
        case InternetConnectionStatus.connected:
          // ignore: avoid_print
          print('Data connection is available.');
          break;
        case InternetConnectionStatus.disconnected:
          // ignore: avoid_print
          print('You are disconnected from the internet.');
          break;
      }
    },
  );

  // close listener after 30 seconds, so the program doesn't run forever
  await Future<void>.delayed(const Duration(seconds: 30));
  await listener.cancel();
}


lib/address_check_options.dart
part of internet_connection_checker;

/// This class should be pretty self-explanatory.
/// If [AddressCheckOptions.port]
/// or [AddressCheckOptions.timeout] are not specified, they both
/// default to [InternetConnectionChecker.DEFAULT_PORT]
/// and [InternetConnectionChecker.DEFAULT_TIMEOUT]
/// Also... yeah, I'm not great at naming things.
class AddressCheckOptions {
  /// [AddressCheckOptions] Constructor
  AddressCheckOptions({
    this.address,
    this.hostname,
    this.port = InternetConnectionChecker.DEFAULT_PORT,
    this.timeout = InternetConnectionChecker.DEFAULT_TIMEOUT,
  }) : assert(
    (address != null || hostname != null) && ((address != null) != (hostname != null)),
    'Either address or hostname must be provided, but not both.',
  );

  /// An internet address or a Unix domain address.
  /// This object holds an internet address. If this internet address
  /// is the result of a DNS lookup, the address also holds the hostname
  /// used to make the lookup.
  /// An Internet address combined with a port number represents an
  /// endpoint to which a socket can connect or a listening socket can
  /// bind.
  /// Either [address] or [hostname] must not be null.
  final InternetAddress? address;

  /// The hostname to use for this connection checker.
  /// Will be used to resolve an IP address to open the socket connection to.
  /// Can be used to verify against services that are not guaranteed to have
  /// a fixed IP address. Connecting via hostname also verifies that
  /// DNS resolution is working for the client.
  /// Either [address] or [hostname] must not be null.
  final String? hostname;

  /// Port
  final int port;

  /// Timeout Duration
  final Duration timeout;

  @override
  String toString() => 'AddressCheckOptions($address, $port, $timeout)';
}


lib/internet_connection.dart
part of internet_connection_checker;

/// This is a singleton that can be accessed like a regular constructor
/// i.e. [InternetConnectionChecker()] always returns the same instance.
class InternetConnectionChecker {
  /// This is a singleton that can be accessed like a regular constructor
  /// i.e. InternetConnectionChecker() always returns the same instance.
  factory InternetConnectionChecker() => _instance;

  /// Creates an instance of the [InternetConnectionChecker]. This can be
  /// registered in any dependency injection framework with custom values for
  /// the [checkTimeout] and [checkInterval].
  InternetConnectionChecker.createInstance({
    this.checkTimeout = DEFAULT_TIMEOUT,
    this.checkInterval = DEFAULT_INTERVAL,
    List<AddressCheckOptions>? addresses,
  }) {
    this.addresses = addresses ??
        DEFAULT_ADDRESSES
            .map(
              (AddressCheckOptions e) => AddressCheckOptions(
                address: e.address,
                hostname: e.hostname,
                port: e.port,
                timeout: checkTimeout,
              ),
            )
            .toList();

    // immediately perform an initial check so we know the last status?
    // connectionStatus.then((status) => _lastStatus = status);

    // start sending status updates to onStatusChange when there are listeners
    // (emits only if there's any change since the last status update)
    _statusController.onListen = () {
      _maybeEmitStatusUpdate();
    };

    // stop sending status updates when no one is listening
    _statusController.onCancel = () {
      _timerHandle?.cancel();
      _lastStatus = null; // reset last status
    };
  }

  /// More info on why default port is 53
  /// here:
  /// - https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers
  /// - https://www.google.com/search?q=dns+server+port
  static const int DEFAULT_PORT = 53;

  /// Default timeout is 10 seconds.
  ///
  /// Timeout is the number of seconds before a request is dropped
  /// and an address is considered unreachable
  static const Duration DEFAULT_TIMEOUT = Duration(seconds: 10);

  /// Default interval is 10 seconds
  ///
  /// Interval is the time between automatic checks
  static const Duration DEFAULT_INTERVAL = Duration(seconds: 10);

  /// Predefined reliable addresses. This is opinionated
  /// but should be enough. See https://www.dnsperf.com/#!dns-resolvers
  ///
  /// Addresses info:
  ///
  /// <!-- kinda hackish ^_^ -->
  /// <style>
  /// table {
  ///   width: 100%;
  ///   border-collapse: collapse;
  /// }
  /// th, td { padding: 5px; border: 1px solid lightgrey; }
  /// thead { border-bottom: 2px solid lightgrey; }
  /// </style>
  ///
  /// | Address        | Provider   | Info                                            |
  /// |:---------------|:-----------|:------------------------------------------------|
  /// | 1.1.1.1        | CloudFlare | https://1.1.1.1                                 |
  /// | 1.0.0.1        | CloudFlare | https://1.1.1.1                                 |
  /// | 8.8.8.8        | Google     | https://developers.google.com/speed/public-dns/ |
  /// | 8.8.4.4        | Google     | https://developers.google.com/speed/public-dns/ |
  /// | 208.67.222.222 | OpenDNS    | https://use.opendns.com/                        |
  /// | 208.67.220.220 | OpenDNS    | https://use.opendns.com/                        |
  static final List<AddressCheckOptions> DEFAULT_ADDRESSES =
      List<AddressCheckOptions>.unmodifiable(
    <AddressCheckOptions>[
      AddressCheckOptions(
        address: InternetAddress(
          '1.1.1.1', // CloudFlare
          type: InternetAddressType.IPv4,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2606:4700:4700::1111', // CloudFlare
          type: InternetAddressType.IPv6,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '8.8.4.4', // Google
          type: InternetAddressType.IPv4,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2001:4860:4860::8888', // Google
          type: InternetAddressType.IPv6,
        ),
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '208.67.222.222', // OpenDNS
          type: InternetAddressType.IPv4,
        ), // OpenDNS
      ),
      AddressCheckOptions(
        address: InternetAddress(
          '2620:0:ccc::2', // OpenDNS
          type: InternetAddressType.IPv6,
        ), // OpenDNS
      ),
    ],
  );

  late List<AddressCheckOptions> _addresses;

  /// A list of internet addresses (with port and timeout) to ping.
  ///
  /// These should be globally available destinations.
  /// Default is [DEFAULT_ADDRESSES].
  ///
  /// When [hasConnection] or [connectionStatus] is called,
  /// this utility class tries to ping every address in this list.
  ///
  /// The provided addresses should be good enough to test for data connection
  /// but you can, of course, supply your own.
  ///
  /// See [AddressCheckOptions] for more info.
  List<AddressCheckOptions> get addresses => _addresses;

  set addresses(List<AddressCheckOptions> value) {
    _addresses = List<AddressCheckOptions>.unmodifiable(value);
    _maybeEmitStatusUpdate();
  }

  static final InternetConnectionChecker _instance =
      InternetConnectionChecker.createInstance();

  /// Ping a single address. See [AddressCheckOptions] for
  /// info on the accepted argument.
  Future<AddressCheckResult> isHostReachable(
    AddressCheckOptions options,
  ) async {
    Socket? sock;
    try {
      sock = await Socket.connect(
        // If address is null, the [AddressCheckOptions] constructor will have
        // asserted that hostname must not be null.
        options.address ?? options.hostname,
        options.port,
        timeout: options.timeout,
      )
      ..destroy();
      return AddressCheckResult(
        options,
        isSuccess: true,
      );
    } catch (e) {
      sock?.destroy();
      return AddressCheckResult(
        options,
        isSuccess: false,
      );
    }
  }

  /// Initiates a request to each address in [addresses].
  /// If at least one of the addresses is reachable
  /// we assume an internet connection is available and return `true`.
  /// `false` otherwise.
  Future<bool> get hasConnection async {
    final Completer<bool> result = Completer<bool>();
    int length = addresses.length;

    for (final AddressCheckOptions addressOptions in addresses) {
      // ignore: unawaited_futures
      isHostReachable(addressOptions).then(
        (AddressCheckResult request) {
          length -= 1;
          if (!result.isCompleted) {
            if (request.isSuccess) {
              result.complete(true);
            } else if (length == 0) {
              result.complete(false);
            }
          }
        },
      );
    }

    return result.future;
  }

  /// Initiates a request to each address in [addresses].
  /// If at least one of the addresses is reachable
  /// we assume an internet connection is available and return
  /// [InternetConnectionStatus.connected].
  /// [InternetConnectionStatus.disconnected] otherwise.
  Future<InternetConnectionStatus> get connectionStatus async {
    return await hasConnection
        ? InternetConnectionStatus.connected
        : InternetConnectionStatus.disconnected;
  }

  /// The interval between periodic checks. Periodic checks are
  /// only made if there's an attached listener to [onStatusChange].
  /// If that's the case [onStatusChange] emits an update only if
  /// there's change from the previous status.
  ///
  /// Defaults to [DEFAULT_INTERVAL] (10 seconds).
  final Duration checkInterval;

  /// The timeout period before a check request is dropped and an address is
  /// considered unreachable.
  ///
  /// Defaults to [DEFAULT_TIMEOUT] (10 seconds).
  final Duration checkTimeout;

  // Checks the current status, compares it with the last and emits
  // an event only if there's a change and there are attached listeners
  //
  // If there are listeners, a timer is started which runs this function again
  // after the specified time in 'checkInterval'
  Future<void> _maybeEmitStatusUpdate([
    Timer? timer,
  ]) async {
    // just in case
    _timerHandle?.cancel();
    timer?.cancel();

    final InternetConnectionStatus currentStatus = await connectionStatus;

    // only send status update if last status differs from current
    // and if someone is actually listening
    if (_lastStatus != currentStatus && _statusController.hasListener) {
      _statusController.add(currentStatus);
    }

    // start new timer only if there are listeners
    if (!_statusController.hasListener) return;
    _timerHandle = Timer(checkInterval, _maybeEmitStatusUpdate);

    // update last status
    _lastStatus = currentStatus;
  }

  // _lastStatus should only be set by _maybeEmitStatusUpdate()
  // and the _statusController's.onCancel event handler
  InternetConnectionStatus? _lastStatus;
  Timer? _timerHandle;

  // controller for the exposed 'onStatusChange' Stream
  final StreamController<InternetConnectionStatus> _statusController =
      StreamController<InternetConnectionStatus>.broadcast();

  /// Subscribe to this stream to receive events whenever the
  /// [InternetConnectionStatus] changes. When a listener is attached
  /// a check is performed immediately and the status ([InternetConnectionStatus])
  /// is emitted. After that a timer starts which performs
  /// checks with the specified interval - [checkInterval].
  /// Default is [DEFAULT_INTERVAL].
  ///
  /// *As long as there's an attached listener, checks are being performed,
  /// so remember to dispose of the subscriptions when they're no longer needed.*
  ///
  /// Example:
  ///
  /// ```dart
  /// var listener = InternetConnectionChecker().onStatusChange.listen((status) {
  ///   switch(status) {
  ///     case InternetConnectionStatus.connected:
  ///       print('Data connection is available.');
  ///       break;
  ///     case InternetConnectionStatus.disconnected:
  ///       print('You are disconnected from the internet.');
  ///       break;
  ///   }
  /// });
  /// ```
  ///
  /// *Note: Remember to dispose of any listeners,
  /// when they're not needed anymore,
  /// e.g. in a* `StatefulWidget`'s *dispose() method*
  ///
  /// ```dart
  /// ...
  /// @override
  /// void dispose() {
  ///   listener.cancel();
  ///   super.dispose();
  /// }
  /// ...
  /// ```
  ///
  /// For as long as there's an attached listener, requests are
  /// being made with an interval of `checkInterval`. The timer stops
  /// when an automatic check is currently executed, so this interval
  /// is a bit longer actually (the maximum would be `checkInterval` +
  /// the maximum timeout for an address in `addresses`). This is by design
  /// to prevent multiple automatic calls to `connectionStatus`, which
  /// would wreck havoc.
  ///
  /// You can, of course, override this behavior by implementing your own
  /// variation of time-based checks and calling either `connectionStatus`
  /// or `hasConnection` as many times as you want.
  ///
  /// When all the listeners are removed from `onStatusChange`, the internal
  /// timer is cancelled and the stream does not emit events.
  Stream<InternetConnectionStatus> get onStatusChange =>
      _statusController.stream;

  /// Returns true if there are any listeners attached to [onStatusChange]
  bool get hasListeners => _statusController.hasListener;

  /// Alias for [hasListeners]
  bool get isActivelyChecking => _statusController.hasListener;
}
